### PR TITLE
[wasm] Enable System.Threading.Tasks.Extensions tests

### PR DESCRIPTION
**Pull request by mdh1418**
Contributes to #38422 
`Tests run: 184, Errors: 0, Failures: 0, Skipped: 1. Time: 0.350004s`

### CODE REVIEW

**Review by akoeplinger — APPROVED** [review]

**Comment by stephentoub** [comment]
> The following 9 tests all failed with System.PlatformNotSupportedException : Cannot wait on monitors on this runtime.

Most of these can easily be tweaked to avoid needing that.

**Comment by steveisok** [comment]
@stephentoub What would you suggest?

**Comment by stephentoub** [comment]
> @stephentoub What would you suggest?

Like this:
https://github.com/dotnet/runtime/pull/38815/commits/13fe8ee609abdd232be6ba9a1a6d7fff3ef236a1

**Comment by stephentoub** [comment]
This test suite still has a bunch of conditional checks for IsThreadingSupported... are all of those still needed?  Or are they leftover from before `async Task` tests worked?  If the latter, can we clean that up here and anywhere else we may have created debt?  Thanks.

**Comment by mdh1418** [comment]
Some new tests were failing locally for me, so I changed them to `ConditionalFact`
- `System.Threading.Tasks.Sources.Tests.ManualResetValueTaskSourceTests.OnCompleted_ContinuationAlwaysInvokedAsynchronously`
- `System.Threading.Tasks.Sources.Tests.ManualResetValueTaskSourceTests.SetResult_RunContinuationsAsynchronously_ContinuationInvokedAccordingly`

Both of these had `tp callback threw a Xunit.Sdk.NotEqualException`.

The other instances of `IsThreadingSupported` are necessary because the tests otherwise hang. I reran for each of the 23 skipped tests, trying with `[Fact]` or `[Theory]` instead, and ran for each of the test cases.

**Comment by akoeplinger** [comment]
> Both of these had tp callback threw a Xunit.Sdk.NotEqualException.

That's expected since the test does `Assert.NotEqual(threadId, Environment.CurrentManagedThreadId);` to verify there's a different thread, which won't be true in a single-thread environment.

Side note: I wonder whether this should've actually triggered the `AppDomain.UnhandledException` instead of printing this message: https://github.com/dotnet/runtime/blob/af4f8c240baa547485fcd0c4d6ae6530d3c7e473/src/mono/mono/mini/mini-wasm.c#L611-L614

**Comment by stephentoub** [comment]
> That's expected since the test does Assert.NotEqual(threadId, Environment.CurrentManagedThreadId); to verify there's a different thread, which won't be true in a single-thread environment.

The goal of the tests isn't to validate it's a different thread, just that the continuation is invoked asynchronously rather than synchronously.  We can validate that in other ways.  For example, we can store a value into an `ThreadLocal<T>` and set some sentinel value in it, then in the callback validate that the `ThreadLocal<T>` does or does not contain the desired value, and then remove the value from the `ThreadLocal<T>` after the line that does the validation, e.g.
```C#
var tl = new ThreadLocal<T> { Value = 42;
InvokeAsynchronously(() => Assert.NotEqual(42, tl.Value));
tl.Value = 0;
```

In general I'd like disabling of such tests to be a last resort.

File: src/libraries/System.Threading.Tasks.Extensions/tests/AsyncValueTaskMethodBuilderTests.cs [hunk 1]
```diff
@@ -399,7 +399,7 @@ async ValueTask<int> ValueTaskReturningAsyncMethod(int result)
             }
         }
 
-        [Fact]
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
```
Comment (stephentoub) [inline src/libraries/System.Threading.Tasks.Extensions/tests/AsyncValueTaskMethodBuilderTests.cs:402]: If the test is hanging, shouldn't that be an ActiveIssue?

**Comment by akoeplinger** [comment]
> The goal of the tests isn't to validate it's a different thread, just that the continuation is invoked asynchronously rather than synchronously

Yeah I meant that was what the test is doing right now, we can obviously rewrite it :)

File: src/libraries/System.Threading.Tasks.Extensions/tests/AsyncValueTaskMethodBuilderTests.cs (hunk 1 above)
Comment (akoeplinger) [reply src/libraries/System.Threading.Tasks.Extensions/tests/AsyncValueTaskMethodBuilderTests.cs:402]: I've looked at it and it hangs in the threadpool pump loop in xharness as soon as we hit the "Incomplete" tests. I didn't have time to investigate more deeply but I agree, we should file an ActiveIssue for these cases.

Comment (mdh1418) [reply src/libraries/System.Threading.Tasks.Extensions/tests/AsyncValueTaskMethodBuilderTests.cs]: I created an ActiveIssue here https://github.com/dotnet/runtime/issues/38931

**Comment by mdh1418** [comment]
The failing lane is in Linux x64, which is unrelated to changes in System.Threading.Tasks.Extensions for wasm.

**Pull request merged**